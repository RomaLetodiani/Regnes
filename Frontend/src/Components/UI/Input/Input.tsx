import { InputHTMLAttributes } from "react";
import { twMerge } from "tailwind-merge";

export interface inputI extends InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  labelClassName?: string;
  inputClassName?: string;
  wrapperClassName?: string;
  errorMessage?: string;
  hasError?: boolean;
  focus?: boolean;
}

const Input = ({
  label,
  labelClassName,
  inputClassName,
  wrapperClassName,
  errorMessage,
  hasError,
  focus,
  ...rest
}: inputI) => {
  return (
    <div className={twMerge("relative w-full", wrapperClassName)}>
      <label
        htmlFor={rest.id}
        className={twMerge(
          "absolute left-3 pointer-events-none transition-all",
          `${
            focus
              ? "top-1 text-sm text-primary"
              : "transform top-1/2 -translate-y-1/2 text-secondary-200"
          }`,
          hasError && "text-danger",
          labelClassName
        )}
      >
        {hasError ? errorMessage : label}
      </label>
      <input
        {...rest}
        className={twMerge(
          "w-full border outline-none px-3 py-2 rounded-xl",
          label && "pt-5",
          `${hasError && "border-danger"}`,
          inputClassName
        )}
      />
    </div>
  );
};

export default Input;
