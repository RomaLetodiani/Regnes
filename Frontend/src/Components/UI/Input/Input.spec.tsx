import { render, screen, cleanup } from "@testing-library/react";
import Input, { inputI } from "./Input";

// Helper function to render the Input component
const renderInput = (props: inputI) => {
  render(<Input {...props} />);
};

describe("Input", () => {
  afterEach(cleanup);

  test("renders basic input with label and id", () => {
    renderInput({ label: "Email", id: "email" });

    const input = screen.getByLabelText("Email");
    const label = screen.getByText("Email");
    expect(input).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(input).toHaveAttribute("id", "email");
  });

  test("label moves up on focus", async () => {
    renderInput({ label: "Email", id: "email", focus: true });
    const label = screen.getByText("Email");

    expect(label).toHaveClass("top-1");
  });

  test("renders error message when hasError", () => {
    renderInput({ label: "Email", id: "email", hasError: true, errorMessage: "Email is required" });
    const errorMessage = screen.getByText("Email is required");

    expect(errorMessage).toBeInTheDocument();
  });

  test("input is disabled when disabled prop is true", () => {
    renderInput({ label: "Email", id: "email", disabled: true });
    const input = screen.getByLabelText("Email");

    expect(input).toBeDisabled();
  });

  test("renders different input types correctly", () => {
    renderInput({ label: "Password", id: "password", type: "password" });
    const input = screen.getByLabelText("Password");

    expect(input).toHaveAttribute("type", "password");
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<Input label="Email" id="email" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
