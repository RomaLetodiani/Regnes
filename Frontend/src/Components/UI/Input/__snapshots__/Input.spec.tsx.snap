// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative w-full"
  >
    <label
      class="absolute left-3 pointer-events-none transition-all transform top-1/2 -translate-y-1/2 text-secondary-200"
      for="email"
    >
      Email
    </label>
    <input
      class="w-full border outline-none px-3 py-2 rounded-xl pt-5 undefined"
      id="email"
    />
  </div>
</DocumentFragment>
`;
